pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go

```diff
@@ -168,13 +200,45 @@ func (plugin *kubenetNetworkPlugin) Init(host network.Host, hairpinMode kubeletc
 
 // TODO: move thic logic into cni bridge plugin and remove this from kubenet
 func (plugin *kubenetNetworkPlugin) ensureMasqRule() error {
-	if plugin.nonMasqueradeCIDR != "0.0.0.0/0" {
-		if _, err := plugin.iptables.EnsureRule(utiliptables.Append, utiliptables.TableNAT, utiliptables.ChainPostrouting,
-			"-m", "comment", "--comment", "kubenet: SNAT for outbound traffic from cluster",
-			"-m", "addrtype", "!", "--dst-type", "LOCAL",
-			"!", "-d", plugin.nonMasqueradeCIDR,
-			"-j", "MASQUERADE"); err != nil {
-			return fmt.Errorf("Failed to ensure that %s chain %s jumps to MASQUERADE: %v", utiliptables.TableNAT, utiliptables.ChainPostrouting, err)
+	for _, currentCIDR := range plugin.nonMasqueradeCIDRs {
+		if currentCIDR != zeroCIDRv4 && currentCIDR != zeroCIDRv6 {
+			// switch according to target nonMasqueradeCidr ip family
+			ipt := plugin.iptables
+			if netutils.IsIPv6CIDRString(currentCIDR) {
+				ipt = plugin.iptablesv6
+			}
+
+			if _, err := ipt.EnsureRule(utiliptables.Append, utiliptables.TableNAT, utiliptables.ChainPostrouting,
```

Comment:
I think the logic here is buggy.  This code needs to emit one rule for each IP version (4 and 6), where that rule says to masquerade if the destination address is not LOCAL and is in NONE OF the version's nonMasqueradeCIDRs.  The existing code emits one rule per nonMasqueradeCIDRs, which is wrong if there is more than one nonMasqueradeCIDRs for a given IP version.

Reply:
mostly this is actually turned off. Even when routing outside the cluster in the same vnet as the cluster (as long as routing table is the same across). It is a good improvement for a follow up.

Reply:
I am uncomfortable shipping something that will silently do the wrong thing if users violate an unwritten expectation.  If we do not handle multiple nonMasqueradeCIDRs in an IP version now, could we at least document that?